Mathlib/Data/Set/FiniteExhaustion.lean

```diff
@@ -0,0 +1,100 @@
+/-
+Copyright (c) 2025 David Ledvinka. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Ledvinka
+-/
+module
+
+public import Mathlib.Data.Set.Countable
+public import Mathlib.Data.Finite.Prod
+
+/-!
+# Finite Exhaustions
+
+This file defines a structure called `FiniteExhaustion` which represents an exhaustion of a
+countable set by an increasing sequence of finite sets. Given a countable set `s`,
+`FiniteExhaustion.choice s` is a choice of a finite exhaustion.
+-/
+
+@[expose] public section
+
+open Set
+
+/-- A `FiniteExhaustion` of a set `s` is a monotonically increasing sequence
+of finite sets such that their union is `s`. -/
+structure FiniteExhaustion {α : Type*} (s : Set α) where
+  /-- The underlying sequence of a `FiniteExhaustion`. -/
+  toFun : ℕ → Set α
+  /-- Every set in a `FiniteExhaustion` is finite. -/
+  finite' : ∀ n, Finite (toFun n)
+  /-- The sequence of sets in a `FiniteExhaustion` are monotonically increasing. -/
+  subset_succ' : ∀ n, toFun n ⊆ toFun (n + 1)
+  /-- The union of all sets in a `FiniteExhaustion` equals `s` -/
+  iUnion_eq' : ⋃ n, toFun n = s
+
+namespace FiniteExhaustion
+
+instance {α : Type*} {s : Set α} : FunLike (FiniteExhaustion s) ℕ (Set α) where
+  coe := toFun
+  coe_injective' | ⟨_, _, _, _⟩, ⟨_, _, _, _⟩, rfl => rfl
+
+instance {α : Type*} {s : Set α} : RelHomClass (FiniteExhaustion s) LE.le HasSubset.Subset where
+  map_rel K _ _ h := monotone_nat_of_le_succ (fun n ↦ K.subset_succ' n) h
+
+instance {α : Type*} {s : Set α} {K : FiniteExhaustion s} {n : ℕ} : Finite (K n) :=
+  K.finite' n
+
+variable {α : Type*} {s : Set α} (K : FiniteExhaustion s)
+
+@[simp]
+theorem toFun_eq_coe : K.toFun = K := rfl
+
+protected theorem finite (n : ℕ) : (K n).Finite := K.finite' n
+
+theorem subset_succ (n : ℕ) : K n ⊆ K (n + 1) := K.subset_succ' n
+
+protected theorem mono {m n : ℕ} (h : m ≤ n) : K m ⊆ K n :=
+  OrderHomClass.mono K h
+
+theorem iUnion_eq : ⋃ n, K n = s :=
+  K.iUnion_eq'
+
+/-- A choice of a `FiniteExhaustion` for a countable set `s`. -/
+noncomputable def choice (s : Set α) [Countable s] : FiniteExhaustion s := by
```

Comment:
I suspect it will be slightly more convenient to use [Set.Countable](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Set/Countable.html#Set.Countable) rather than [Countable](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Countable/Defs.html#Countable) and given this, for dot notation I'd name this as follows:
```suggestion
noncomputable def _root_.Set.Countable.finiteExhaustion {s : Set α} (hs : s.Countable) :
    FiniteExhaustion s := by
```